Tensile/KernelWriterAssembly.py

```diff
@@ -7196,6 +7222,117 @@ def calculateLdsWriteOffset(self, perp, para, sPerp, sPara, kernel, tP, localWri
+    BranchMod.addInst("s_cmp_eq_u32", sgpr("PreLoopLWVmcntCase"), hex(2), "Case 2: Prev PK-Loop is Opt-NLL?")
+    BranchMod.addInst("s_cbranch_scc1", optNLL_lw_Label, "jump to Case 2")
+    BranchMod.addInst("s_cmp_eq_u32", sgpr("PreLoopLWVmcntCase"), hex(3), "Case 3: Prev PK-Loop is Ord-NLL with no beta?")
+    BranchMod.addInst("s_cbranch_scc1", ordNLL_B0_lw_Label, "jump to Case 3")
+    BranchMod.addInst("s_cmp_eq_u32", sgpr("PreLoopLWVmcntCase"), hex(4), "Case 4: Prev PK-Loop is Ord-NLL with beta?")
+    BranchMod.addInst("s_cbranch_scc1", ordNLL_B1_lw_Label, "jump to Case 4")
+
+    # Fast duplicate LWDoCodeTemplate four times to different placeholder keywords for later replacement (after global write)
+    # can avoid calling localWriteDo() for several times
+
+    # CASE 1:
+    # duplicate vmcnt("__placeholder__ + Basic_GL_vmcnt - Decrement") to vmcnt("Basic_GL_vmcnt - Decrement")
+    LWDoCase1Mod = imod.addCode(Code.Module("PreLoopLW_Case1"))
+    LWDoCase1Mod.addComment1("global-load-cnt = Basic_GL_vmcnt")
+    for item in codeTemplateStrList:
+      LWDoCase1Mod.addText(str(item).replace("__placeholder__+",""))
+    LWDoCase1Mod.addInst("s_branch", lwEnd_Label, "finish case 1, jump to end of LW")
+
+    # CASE 2:
+    # duplicate vmcnt("__placeholder__ + Basic_GL_vmcnt - Decrement") to vmcnt("OptNLL_GW_vmcnt + Basic_GL_vmcnt - Decrement")
+    LWDoCase2Mod = imod.addCode(Code.Module("PreLoopLW_Case2"))
+    LWDoCase2Mod.addText("\n%s:" % optNLL_lw_Label)
+    LWDoCase2Mod.addComment1("prev-global-store-cnt = OptNLL_GW_vmcnt, global-load-cnt = Basic_GL_vmcnt")
+    for item in codeTemplateStrList:
+      LWDoCase2Mod.addText(str(item).replace("__placeholder__","OptNLL_GW_vmcnt"))
+    LWDoCase2Mod.addInst("s_branch", lwEnd_Label, "finish case 2, jump to end of LW")
+
+    # CASE 3:
+    # duplicate vmcnt("__placeholder__ + Basic_GL_vmcnt - Decrement") to vmcnt("OrdNLL_B0_GW_vmcnt + Basic_GL_vmcnt - Decrement")
+    LWDoCase3Mod = imod.addCode(Code.Module("PreLoopLW_Case3"))
+    LWDoCase3Mod.addText("\n%s:" % ordNLL_B0_lw_Label)
+    LWDoCase3Mod.addComment1("prev-global-store-cnt = OrdNLL_B0_GW_vmcnt, global-load-cnt = Basic_GL_vmcnt")
+    for item in codeTemplateStrList:
+      LWDoCase3Mod.addText(str(item).replace("__placeholder__","OrdNLL_B0_GW_vmcnt"))
+    LWDoCase3Mod.addInst("s_branch", lwEnd_Label, "finish case 3, jump to end of LW")
+
+    # CASE 4:
+    # duplicate vmcnt("__placeholder__ + Basic_GL_vmcnt - Decrement") to vmcnt("OrdNLL_B1_GW_vmcnt + Basic_GL_vmcnt - Decrement")
+    LWDoCase4Mod = imod.addCode(Code.Module("PreLoopLW_Case4"))
+    LWDoCase4Mod.addText("\n%s:" % ordNLL_B1_lw_Label)
+    LWDoCase4Mod.addComment1("prev-global-store-cnt = OrdNLL_B1_GW_vmcnt, global-load-cnt = Basic_GL_vmcnt")
+    if kernel["StoreRemapVectorWidth"]:
+      LWDoCase4Mod.addText("\ns_barrier // since no need for any vmcnt for case4, so need a barrier to ensure ds_read of SR is done\n")
```

Comment:
I think all cases, except for first time entering the preloop, needs this barrier when store remapping is enabled. One wave could get here first while other waves are still reading from the LDS.

Reply:
Good catch! Updated and refined